### ISSUE TITLE
Try using a trie or sorted lookup datastructure to improve performance

### ISSUE THREAD

**Issue by aarondandy** [aarondandy/WeCantSpell.Hunspell#22]
During the initial developement of this port, a regular old dictionary of type `Dictionary<string, T>` was used to store various words, affixes, and their associated details. This really helped speed up the development of the code initially and even has some performance strengths in specific cases.

It is clear to me now that the choice to use `Dictionary<,>` was probably wrong for a lot of things as it is not a very good data structure when you want multiple related results for a query. When profiled this is where the code spends most of its time.

I would like to see how the use of a [trie](https://en.wikipedia.org/wiki/Trie) or a sorted collection impacts performance in the library. Looking back at the Hunspell source and some e-mails discussing design I think the original source uses something like a sorted linked list as the main storage for root words.

The following locations in the code may benefit from swapping out and utilizing a better datastructure:

## GetMatchingAffixes

The methods `SuffixCollection.GetMatchingAffixes` and `PrefixCollection.GetMatchingAffixes` both look for affixes that begin or "end" with a certain string of text. This part of the code could probably benefit greatly from having a list of `Affix<>` that can be indexed by word instead of having all `AffixEntries` being confusingly nested into groups. It may also create a case for the internal `Affix<>` type to be converted to a reference type. The `GetMatchingWithDotAffixes` method is also realated but is often a cold path so optimiazation there should be focused on code size.

## FindLargestMatchingConversion

The method `MultiReplacementTable.FindLargestMatchingConversion` while small may make many calls to a dictionary as it is a loop that is itself called from within a loop. It's responsibility is to find the longest matching entry for a substring.

## PatternSet Check

The method `PatternSet.Check` searches for a pattern entry that has a text value that is ia subset of another.

## WordList

While I'm not sure that a trie or even a sorted list would have an impact on the `WordList.EntriesByRoot` collection, the entries beind sorted may have a beneficial impact of keeping related roots near eachother in memory.

**Comment by aarondandy** [comment]
The trie implementation should probably be implemented as read only to optimizew for search. Building it up from roots would be best done by converting a dictionary (in the wordlist.builder)  to a read only trie instance.

**Comment by aarondandy** [comment]
Whatever is chosen ... or made ... will need to support Span<char> as well.

**Comment by aarondandy** [comment]
I tried the trie, it did not go well sadly. It seems that for the workload a hash lookup is still superior. While you would think it would be a help with things like the affix queries, they only measure a small percentage when profiling. This cause the overall performance to become even worse. I also attempted to use a Dictionary that was keyed on a ReadOnlyMemory<char> to reduce the number of .ToString calls used to perform a lookup but the hashing and equality costs of needing to override the comparer did more harm than good. The only solution that may make it better is possibly a sorted list, but again most of the queries are for a specific key and would benefit from the hash. I don't think there is much that can be done here.
